edocumentation-service/src/main/java/org/dcsa/edocumentation/service/unofficial/BookingValidationService.java

```diff
@@ -71,11 +72,90 @@ private String validateBooking(Booking booking) {
     if (isInThePast(today, booking.getExpectedArrivalAtPlaceOfDeliveryEndDate())) {
       return "expectedArrivalAtPlaceOfDeliveryEndDate is in the past";
     }
-
-    return null;
+    return validateShipmentLocations(booking.getShipmentLocations());
   }
 
   private boolean isInThePast(LocalDate today, LocalDate time) {
     return time != null && today.isAfter(time);
   }
+
+  private String validateShipmentLocations(Set<ShipmentLocation> shipmentLocations) {
+    if (shipmentLocations.isEmpty()) {
+      return "Invalid booking: Shipment locations should not be empty";
+    }
+
+    Predicate<LocationType> predicateHasPREorPOL =
+        code -> code == LocationType.PRE || code == LocationType.POL;
+
+    Predicate<LocationType> predicateHasPODorPDE =
+        code -> code == LocationType.POD || code == LocationType.PDE;
+
+    boolean hasPREorPOLs =
+        shipmentLocations.stream()
+            .map(ShipmentLocation::getShipmentLocationTypeCode)
+            .anyMatch(predicateHasPREorPOL);
+
+    if (!hasPREorPOLs) {
+      return "No ShipmentLocationTypeCode of PRE or POL found in the shipmentLocations."
+          + " At least one of them should be provided.";
+    }
+
+    boolean hasPODorPDE =
+        shipmentLocations.stream()
+            .map(ShipmentLocation::getShipmentLocationTypeCode)
+            .anyMatch(predicateHasPODorPDE);
+
+    if (!hasPODorPDE) {
+      return "No ShipmentLocationTypeCode of POD or PDE found in the shipmentLocations."
+          + "At least one of them should be provided.";
+    }
+
+    Predicate<ShipmentLocation> predicathasPREorPOLorPODorPDE =
+        sl ->
+            predicateHasPREorPOL.test(sl.getShipmentLocationTypeCode())
+                && predicateHasPODorPDE.test(sl.getShipmentLocationTypeCode());
+
```

Comment:
The error was in the logic. Since, as you say it always returns False the dependent UN location & address validations were checked against an empty list.  However, I added a case for UN locations too (there is a test for matching addresses).